cmd/libsnap-confine-private/mountinfo.c

```diff
@@ -127,42 +128,77 @@ static void show_buffers(const char *line, int offset,
 	fprintf(stderr, ">%s<\n", line);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i) {
+	for (size_t i = 0; i < strlen(line); ++i) {
 		int c = entry->line_buf[i];
 		fputc(c == 0 ? '@' : c == 1 ? '#' : c, stderr);
 	}
 	fputc('<', stderr);
 	fputc('\n', stderr);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i)
+	for (size_t i = 0; i < strlen(line); ++i)
 		fputc('=', stderr);
 	fputc('<', stderr);
 	fputc('\n', stderr);
 #endif				// MOUNTINFO_DEBUG
 }
 
+static bool isoctal(char c)
+{
+	return c >= '0' && c <= '7';
+}
+
 static char *parse_next_string_field(struct sc_mountinfo_entry *entry,
-				     const char *line, int *offset)
+				     const char *line, size_t * offset)
 {
-	int offset_delta = 0;
-	char *field = &entry->line_buf[0] + *offset;
-	if (line[*offset] == ' ') {
-		// Special case for empty fields which cannot be parsed with %s.
-		*field = '\0';
-		*offset += 1;
-	} else {
-		int nscanned =
-		    sscanf(line + *offset, "%s%n", field, &offset_delta);
-		if (nscanned != 1)
-			return NULL;
-		*offset += offset_delta;
-		if (line[*offset] == ' ') {
-			*offset += 1;
+	const char *input = &line[*offset];
+	char *output = &entry->line_buf[*offset];
+	const char *s;		// Used for escape logic.
+	size_t input_idx = 0;
+	size_t output_idx = 0;
+
+	// Scan characters until we run out of memory to scan or we find a space.
+	// The kernel uses simple octal escape sequences for the following:
+	// space, tab, newline, backwards slash. Everything else is copied
+	// verbatim.
+	bool stop = false;
+	for (int c; c = input[input_idx], c != '\0' && !stop;
+	     input_idx++, output_idx++) {
+		switch (c) {
+		case ' ':
+			// Fields are space delimited. Represent the space
+			// as the end-of-string marker, skip over it, and stop parsing.
+			output[output_idx] = '\0';
+			stop = true;
+			break;
+		case '\\':
+			// Three *more* octal digits required for the escape sequence.
+			// For reference see mangle_path() in fs/seq_file.c.
+			s = &input[input_idx];
+			if (isoctal(s[1]) && isoctal(s[2]) && isoctal(s[3])) {
```

Comment:
This looks slightly tricky and at first I was about to comment it may go out of input boundary on partial escape sequence at the end of the input, but it's actually fine because of null-termination. Perhaps merits a short comment, but maybe it's just me 😊